### PR TITLE
Test 82 Implement Kanban Board

### PR DESCRIPTION
**Pull request by stoyanov-ivelin**
Implemented a Kanban Board with drag&drop functionality.

### CODE REVIEW

File: kanban-board/src/store/store.ts [hunk 1]
```diff
@@ -1,18 +1,36 @@
...
+    case 'DRAG_BEGIN':
+      return {
+        ...state,
+        currentDraggedElement: action.payload
+      }
+    case 'UPDATE_STATUS':
+      const index = state.currentDraggedElement;
+      const newState = { ...state, currentDraggedElement: null };
+      const newStatus = action.payload;
+      newState.issues[index!].status = newStatus;
```
Comment (MilenIvanovv) [inline kanban-board/src/store/store.ts:25]: Here the **!** is a kind of a cheat to stop the compiler. If the  **UPDATE_STATUS** action is used with `state.currentDraggedElement = null` an error will occur.

File: kanban-board/src/store/store.ts [hunk 2]
```diff
@@ -1,18 +1,36 @@
...
-const reducer = (state = initialState, action: any) => {
-  console.log(action);
+const reducer = (state = initialState, action: AnyAction) => {
+  switch(action.type) {
+    case 'DRAG_BEGIN':
+      return {
+        ...state,
+        currentDraggedElement: action.payload
+      }
+    case 'UPDATE_STATUS':
```
Comment (MilenIvanovv) [inline kanban-board/src/store/store.ts:21]: Reuse the constant **UPDATE_STATUS** here.

File: kanban-board/src/components/Issue/Issue.tsx [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+  id: number,
+  dispatch: AppDispatch,
+}
+
+class Issue extends Component<IIssueProps> {
+  handleDragStart(index: number) {
+    this.props.dispatch(dragBegin(index));
+  }
+
+  render(): ReactNode {
```
Comment (MilenIvanovv) [inline kanban-board/src/components/Issue/Issue.tsx:26]: An abstract component structure we use is:
1. component hooks 
2. render 
3. other methods

File: kanban-board/src/components/Issue/Issue.tsx [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+import { dragBegin } from "../../common/actions";
+import { Status } from "../../common/constants";
+import { IIssue } from "../../common/types";
+import { AppDispatch } from "../../store/store";
+import './Issue.css';
+
+interface IIssueProps {
+  issue: IIssue,
+  id: number,
+  dispatch: AppDispatch,
```
Comment (MilenIvanovv) [inline kanban-board/src/components/Issue/Issue.tsx:18]: This is a very interesting way of connecting the component to the store to me. In Datadash we use **mapDispatchToProp** in **connect()**.

I would have done it one of two ways:
1. connect Issue component to the store
2. dispatch an action in the parent and pass a parent methods

**Review by MilenIvanovv — CHANGES_REQUESTED** [review]
The code looks great to me. Most of the comments I left show how we solve the same problems in a different way, in DataDash

File: kanban-board/src/App.css [hunk 5]
```diff
@@ -2,37 +2,8 @@
...
-
-@keyframes App-logo-spin {
-  from {
-    transform: rotate(0deg);
-  }
-  to {
-    transform: rotate(360deg);
-  }
-}
+.columnGrid {
```
[the remark this answers is not in the document]
Comment (stoyanov-ivelin) [reply kanban-board/src/App.css]: done

File: kanban-board/src/common/constants.ts [hunk 6]
```diff
@@ -0,0 +1,5 @@
+export enum Status {
```
[the remark this answers is not in the document]
Comment (stoyanov-ivelin) [reply kanban-board/src/common/constants.ts:1]: done

File: kanban-board/src/components/Board/Board.tsx [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+  }
+
+  render(): ReactNode {
+    return (
+      <Grid container justifyContent="center" spacing={2}>
+        <Grid item>
+          <Paper
+            className="paper"
+            style={{backgroundColor: "lightgrey"}}
+            onDragEnter={(e) => this.handleDragEnter(e)}
```
[the remark this answers is not in the document]
Comment (stoyanov-ivelin) [reply kanban-board/src/components/Board/Board.tsx:25]: Aight, will keep it in mind

File: kanban-board/src/App.tsx [hunk 8]
```diff
@@ -1,18 +1,74 @@
...
+    const doneIssues = filteredIssues.done;
+
     return (
-    <div className="App">
-      Hello, World!
-    </div>
-    )};
+      <div className="App">
+        <AppBar position="static">
+          <Toolbar style={{ left: "140em" }}>
```
[the remark this answers is not in the document]
Comment (stoyanov-ivelin) [reply kanban-board/src/App.tsx]: done

File: kanban-board/src/App.tsx [hunk 9]
```diff
@@ -1,18 +1,74 @@
...
+            <Typography variant="h5">User Name</Typography>
+            <AccountCircle style={{ marginLeft: "1em" }} />
```
[the remark this answers is not in the document]
Comment (stoyanov-ivelin) [reply kanban-board/src/App.tsx]: done

File: kanban-board/src/store/store.ts (hunk 1 above)
Comment (stoyanov-ivelin) [reply kanban-board/src/store/store.ts]: done, changed the initial value to 0

File: kanban-board/src/store/store.ts (hunk 2 above)
Comment (stoyanov-ivelin) [reply kanban-board/src/store/store.ts]: done

File: kanban-board/src/components/Issue/Issue.tsx (hunk 3 above)
Comment (stoyanov-ivelin) [reply kanban-board/src/components/Issue/Issue.tsx:23]: done

File: kanban-board/src/store/store.ts (hunk 1 above)
Comment (MilenIvanovv) [reply kanban-board/src/store/store.ts]: Why the first issue in the array? What happens if there are no issues?

Comment (MilenIvanovv) [reply kanban-board/src/store/store.ts]: I missed a problem here in my last review.

The action is called **UPDATE_STATUS**. It should not know about the dragging feature.

File: src/store/store.ts [hunk 10]
```diff
@@ -9,23 +9,24 @@ const initialState = {
...
-        currentDraggedElement: action.payload
-      }
     case UPDATE_STATUS:
-      const index = state.currentDraggedElement;
-      const newState = { ...state, currentDraggedElement: null };
-      const newStatus = action.payload;
-      newState.issues[index].status = newStatus;
+      const index = action.payload.issueId;
+      const newStatus = action.payload.newStatus;
+      const newState = state.issues.map(issue => {
```
Comment (MilenIvanovv) [inline src/store/store.ts:19]: I think **find** is a more suitable method here

File: src/components/Board/Board.tsx [hunk 11]
```diff
@@ -23,9 +24,8 @@ class Board extends Component<IBoardProps> {
             className="paper"
             style={{backgroundColor: "lightgrey"}}
             onDragEnter={(e) => this.handleDragEnter(e)}
-            onDrop={(e) => this.handleDrop(e)}
-            onDragEnd={(e) => this.handleDragEnd(e)}
             onDragOver={(e) => this.handleDragOver(e)}
+            onDrop={(e) => this.handleDrop(e)}
```
Comment (MilenIvanovv) [inline src/components/Board/Board.tsx:28]: A small thing, but you can pass directly the function by referece
`onDrop={this.handleDrop}`

File: src/common/types.ts [hunk 12]
```diff
@@ -0,0 +1,14 @@
+export interface IIssue {
+  id: number;
+  name: string;
+  description: string;
+  status: string;
+  assignee: string;
+}
+
+export type UpdateStatusPaylod = {
```
Comment (MilenIvanovv) [inline src/common/types.ts:9]: Use mainly interfaces for abstraction. If you can't then use type. A name for this file I have seen is models.ts or interfaces.ts

File: src/App.tsx [hunk 13]
```diff
@@ -0,0 +1,76 @@
+import { AppBar, Toolbar, Typography } from "@material-ui/core";
+import { Component, ReactNode } from "react";
+import { connect } from "react-redux";
+import "./App.css";
+import { Status } from "./common/constants";
```
Comment (MilenIvanovv) [inline src/App.tsx:5]: Configure the project to use absolute path imports. When the project becomes bigger they are much easier to work with.

**Review by MilenIvanovv — CHANGES_REQUESTED** [review]
Left some comments for minor changes. Great job finding a second solution for saving the dragged element id. Both ways are fine.

I noticed that in the project you are using relative path imports. In DataDash we use absolute path imports. Collaborate with @kristiyanplpeev  to save you some time.

File: src/store/store.ts (hunk 10 above)
Comment (stoyanov-ivelin) [reply src/store/store.ts:19]: done

File: src/components/Board/Board.tsx (hunk 11 above)
Comment (stoyanov-ivelin) [reply src/components/Board/Board.tsx:28]: done

File: src/App.tsx (hunk 13 above)
Comment (stoyanov-ivelin) [reply src/App.tsx:5]: done

**Review by MilenIvanovv — APPROVED** [review]
LGTM

**Pull request merged**